spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapterTests.java

```diff
@@ -126,59 +119,60 @@ public void testAckModeAutoAck() throws Exception {
 				this.pubSubSubscriberOperations, "testSubscription");
 
 		adapter.setAckMode(AckMode.AUTO_ACK);
-
-		when(this.pubSubSubscriberOperations.subscribe(anyString(), any(MessageReceiver.class))).then(this.testAnswer);
-
 		adapter.setOutputChannel(this.messageChannel);
 
+		doAnswer(new CallbackAnswer(this, NACK)).when(
+				this.ackReplyConsumer).nack();
+
 		try {
 			adapter.start();
+
+			Assert.fail(EXPECTED_EXCEPTION);
 		}
 		catch (Throwable t) {
-
+			Assert.assertEquals(EXCEPTION_MESSAGE, t.getCause().getMessage());
 		}
 
-		Assert.assertTrue(this.countDownLatch.await(10, TimeUnit.SECONDS));
-		Assert.assertNull(this.returnValue.get());
+		Assert.assertNull(this.value);
 	}
 
-	private class TestAnswer implements Answer<MessageReceiver> {
+	public void setValue(String value) {
+		this.value = value;
+	}
 
-		private CountDownLatch countDownLatch;
+	private class MessageReceiverAnswer implements Answer<Void> {
 
 		private AckReplyConsumer ackReplyConsumer;
 
-		TestAnswer(CountDownLatch countDownLatch, AckReplyConsumer ackReplyConsumer) {
-			this.countDownLatch = countDownLatch;
+		MessageReceiverAnswer(AckReplyConsumer ackReplyConsumer) {
 			this.ackReplyConsumer = ackReplyConsumer;
 		}
 
-		public MessageReceiver answer(InvocationOnMock invocation) throws Throwable {
-			try {
-				PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(
-						ByteString.copyFrom("Testing 1 2 3".getBytes("UTF-8"))).build();
+		public Void answer(InvocationOnMock invocation) throws Throwable {
+			PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(
+					ByteString.copyFrom(
+							"Testing 1 2 3".getBytes("UTF-8"))).build();
 
-				MessageReceiver messageReceiver = invocation.getArgument(1);
-				messageReceiver.receiveMessage(pubsubMessage, this.ackReplyConsumer);
-			}
-			finally {
-				this.countDownLatch.countDown();
-			}
+			MessageReceiver messageReceiver = invocation.getArgument(1);
+			messageReceiver.receiveMessage(pubsubMessage, this.ackReplyConsumer);
 
 			return null;
 		}
 	}
 
-	private class NackAnswer implements Answer<Boolean> {
+	private class CallbackAnswer implements Answer<Void> {
+
+		private PubSubInboundChannelAdapterTests callback;
```

Comment:
I don't see reason in this inner class because there is a simple access to the `PubSubInboundChannelAdapterTests` as a natural outer-inner classes relationship.
Therefore there is also no need in the `setValue()` because you simple can set it from the lambda:
```
doAnswer(invocation -> {
     this.value = NACK;
     return null;
}).when(this.ackReplyConsumer).nack();
```

That's it. I don't see where is a complexity and overhead during reading.
But with your explicit `PubSubInboundChannelAdapterTests` instance propagation I have to think about its reason and that's confusing. 🤷‍♂️